src/Strategy/AlpineLinux/Parser.hs

```diff
@@ -0,0 +1,128 @@
+-- | Parsing functions for APK files.
+-- APK is the package manager for Alpine Linux (https://alpinelinux.org/).
+--
+-- The database itself is just a flat file with entries in text.  These are
+-- specified in a key-value format with obscure names.  See
+-- https://wiki.alpinelinux.org/wiki/Apk_spec#Syntax.
+module Strategy.AlpineLinux.Parser (installedPackagesDatabaseParser, PackageError (..)) where
+
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.List.NonEmpty qualified as NE
+import Data.Map qualified as Map
+import Data.String.Conversion (toText)
+import Data.Text (Text)
+import Data.Void (Void)
+import Strategy.AlpineLinux.Types (AlpinePackage (..))
+import Text.Megaparsec (
+  MonadParsec (eof, takeWhileP),
+  Parsec,
+  many,
+  optional,
+  some,
+  try,
+ )
+import Text.Megaparsec.Char (char, eol, letterChar)
+
+-- | Errors that can occur even in a valid file.  Packages are a lose collection
+-- of fields, but it unclear which are required.  These error represent fields
+-- we expect, but which may be missing in unusual circumstances.
+data PackageError
+  = MissingPackageName
+  | MissingPackageArchitecture {packageName :: Text}
+  | MissingPackageVersion {packageName :: Text}
+  deriving (Show, Eq, Ord)
+
+type Parser = Parsec Void Text
+
+-- | Key for the package-name field.
+packageNameKey :: Text
+packageNameKey = "P"
+
+-- | Key for the architecture field.
+architectureKey :: Text
+architectureKey = "A"
+
+-- | Key for the version field.
+versionKey :: Text
+versionKey = "V"
+
+-- | Parses zero or more occurrences of a construct with separators between.
+--
+-- This differs from `Control.Monad.Combinators.sepBy` because it will backtrack
+-- failed repetitions.  The generic version has no notion of backtracking so if
+-- the separator consumes input after the last entry then the whole parser will
+-- fail because it will fail to parse another entry after the separator.
+--
+-- TODO: If this is more generally useful we should move it to a common location.
+--
+-- The Megaparsec project chose not to include this in their library:
+--    https://github.com/mrkkrp/megaparsec/issues/401
+-- This implementation is written to mirror the source of the generic version:
+--   https://hackage.haskell.org/package/parser-combinators-1.3.0/docs/src/Control.Monad.Combinators.html#sepBy
+sepByWithTry :: MonadParsec e s m => m a -> m b -> m [a]
+sepByWithTry p sep = do
+  r <- optional (try p)
+  case r of
+    Nothing -> pure []
+    Just x -> (x :) <$> many (try (sep >> p))
```

Comment:
Does `many . try` create an infinite loop?

Reply:
I'm still not sure why we need all of the backtracking logic.  Can you explain why we can't just do `try (sepBy p sep)`?

Reply:
Sure, I'll try to explain it and it'll help me understand it more clearly along the way.

Let's say that we have a input that describes two packages, like this:

```
prop1
prop2

prop3
prop4
```

You could generalize this to a stream of `prop<eol>prop<eol><eol>prop<eol>prop` where `prop` is parsed by whatever parses properties and `eol` is the end-of-line.

`sepBy` works by consuming the first `prop`, then many `eol >> prop` sequences.  On this input that means it will consume the first `prop`, then `<eol>prop`, but then it goes to do it's second `eol >> prop`.  The `eol` succeeds, but `prop` fails.  That actually causes the whole `sepBy` to fail!  If we put `try` around the `sepBy` it will still fail, it'll just backtrack when it does it.  What we want is to backtrack that failing separator that gets consumed at the end.

This Issue sort of explains it too: https://github.com/mrkkrp/megaparsec/issues/401